lib/scanner/walk.go

```diff
@@ -242,40 +274,50 @@ func (w *walker) walkAndHashFiles(ctx context.Context, fchan, dchan chan protoco
 			return skip
 		}
 
-		if w.Matcher.Match(path).IsIgnored() {
-			l.Debugln("ignored (patterns):", path)
-			return skip
-		}
-
 		if !utf8.ValidString(path) {
 			l.Warnf("File name %q is not in UTF8 encoding; skipping.", path)
 			return skip
 		}
 
-		path, shouldSkip := w.normalizePath(path, info)
-		if shouldSkip {
-			return skip
-		}
-
-		switch {
-		case info.IsSymlink():
-			if err := w.walkSymlink(ctx, path, dchan); err != nil {
-				return err
+		if w.Matcher.Match(path).IsIgnored() {
+			l.Debugln("ignored (patterns):", path)
+			if w.Matcher.SkipIgnoredDirs() {
+				return skip
 			}
-			if info.IsDir() {
-				// under no circumstances shall we descend into a symlink
-				return fs.SkipDir
+			// If the parent wasn't ignored already, set this path as the "highest" ignored parent
+			if ignoredParent == "" || !strings.HasPrefix(path, ignoredParent+string(fs.PathSeparator)) {
+				ignoredParent = path
 			}
 			return nil
+		}
 
-		case info.IsDir():
-			err = w.walkDir(ctx, path, info, dchan)
+		if ignoredParent == "" {
+			// parent isn't ignored, nothing special
+			return handleItem(path)
+		}
 
-		case info.IsRegular():
-			err = w.walkRegular(ctx, path, info, fchan)
+		// Part of current path below the ignored (potential) parent
+		rel := strings.TrimPrefix(path, ignoredParent+string(fs.PathSeparator))
+
+		if rel == path {
+			// ignored path isn't actually a parent of the current path
+			ignoredParent = ""
+			return handleItem(path)
 		}
 
-		return err
+		// The previously ignored parent directories of the current, not
+		// ignored path need to be handled as well.
+		if err = handleItem(ignoredParent); err != nil {
+			return err
+		}
+		for _, name := range strings.Split(rel, string(fs.PathSeparator)) {
```

Comment:
Why did the filepath.Base disappear?

Reply:
Because you didn't like it :)  
Before I used it to excluded the current path and handle it at the end of the function. Now I  handle it in this loop and at the end of the function there is just a `return nil` as no error occurred before.

Reply:
I indeed lost the line that prevents this when simplifying the code: After hitting `e`, `ignoredParent` must be reset of course (because the parent isn't ignored anymore).

Reply:
Actually I didn't read your example carefully enough, because I jumped at my mistake: `e` and all the others are not ignored or included, so why should they trigger handling?

Reply:
Well ignoring works by patterns, and my gut says that `/a/b/c/d` expands to something like `/a/b/c/d/**` pattern that then matches everything under that.

Your tests don't handle anything deeper than 1 level it seems.

Reply:
I extended the tests to cover multiple levels.

Reply:
Yeah so I'm also a bit skeptical here. Consider, patterns:

```
!ignore/d1/d2/foo
ignore
```

when we reach `ignore` it will get ignored and set as the parent, remembered. Then we continue walking and at some point reach the unignored `ignore/d1/d2/foo`. We'll start the loop to look at parents, `ignore/d1` to begin with. This is a directory, so it will hit the `walkDir` in the closure above and start walking recursively, at some point reaching `ignore/d1/d2/foo` again, ... ?